internal/pkg/bulk/engine.go

```diff
@@ -98,13 +114,119 @@ func NewBulker(es esapi.Transport, tracer *apm.Tracer, opts ...BulkOpt) *Bulker
 	}
 
 	return &Bulker{
-		opts:        bopts,
-		es:          es,
-		ch:          make(chan *bulkT, bopts.blockQueueSz),
-		blkPool:     sync.Pool{New: poolFunc},
-		apikeyLimit: semaphore.NewWeighted(int64(bopts.apikeyMaxParallel)),
-		tracer:      tracer,
+		opts:                  bopts,
+		es:                    es,
+		ch:                    make(chan *bulkT, bopts.blockQueueSz),
+		blkPool:               sync.Pool{New: poolFunc},
+		apikeyLimit:           semaphore.NewWeighted(int64(bopts.apikeyMaxParallel)),
+		tracer:                tracer,
+		remoteOutputConfigMap: make(map[string]map[string]interface{}),
+		bulkerMap:             make(map[string]Bulk),
+		remoteOutputErrorMap:  make(map[string]string),
+	}
+}
+
+func (b *Bulker) GetRemoteOutputErrorMap() map[string]string {
+	return b.remoteOutputErrorMap
+}
+
+func (b *Bulker) SetRemoteOutputError(name string, status string) {
+	b.remoteOutputErrorMap[name] = status
+}
+
+func (b *Bulker) GetBulker(outputName string) Bulk {
+	return b.bulkerMap[outputName]
+}
+
+func (b *Bulker) GetBulkerMap() map[string]Bulk {
+	return b.bulkerMap
+}
+
+func (b *Bulker) CancelFn() context.CancelFunc {
+	return b.cancelFn
+}
+
+func (b *Bulker) CreateAndGetBulker(zlog zerolog.Logger, outputName string, serviceToken string, outputMap map[string]map[string]interface{}) (Bulk, bool, error) {
+	hasConfigChanged := b.CheckRemoteOutputChanged(zlog, outputName, outputMap[outputName])
+	bulker := b.bulkerMap[outputName]
+	if bulker != nil && !hasConfigChanged {
+		return bulker, false, nil
+	}
+	if bulker != nil && hasConfigChanged {
+		cancelFn := bulker.CancelFn()
+		if cancelFn != nil {
+			cancelFn()
+		}
+	}
+	bulkCtx, bulkCancel := context.WithCancel(context.Background())
```

Comment:
added the logic to do this, how can I test that the remote bulkers are cancelled?

Reply:
I've added a test to verify that the old child bulker ctx is canceled or config change here, but it doesn't seem to happen, not sure what I'm missing, it logs bulk context done, but still tries to make the http request.
[commit](https://github.com/elastic/fleet-server/pull/3051/commits/9382975df36401c4b859516a363de746a52e344f)

```
=== RUN   TestCancelCtxChildBulkerReplaced
    engine.go:261: {"level":"info","name":"remote","message":"remote output configuration has changed"}
    engine.go:194: {"level":"debug","outputName":"remote","message":"Bulk context done"}
    engine.go:183: {"level":"debug","outputName":"remote","message":"Bulker started"}
    engine.go:183: {"level":"debug","outputName":"remote","message":"Bulker started"}
    bulk_test.go:350: apikey auth request : dial tcp: lookup remote-es: no such host
--- PASS: TestCancelCtxChildBulkerReplaced (0.10s)
PASS
ok      github.com/elastic/fleet-server/v7/internal/pkg/bulk    0.502s
```

Reply:
Commented on the test case, i think it's because API key operations don't get processed through bulk calls/the engine so they are not associated with the context created here